datafusion/optimizer/src/eliminate_limit.rs

```diff
@@ -157,180 +80,148 @@ mod tests {
-        let table_scan_inner = test_table_scan_with_name("test1").unwrap();
-        let plan = LogicalPlanBuilder::from(table_scan)
-            .limit(2, Some(1))
-            .unwrap()
+            .limit(2, Some(1))?
             .join_using(
                 &table_scan_inner,
                 JoinType::Inner,
                 vec![Column::from_name("a".to_string())],
-            )
-            .unwrap()
-            .limit(3, Some(1))
-            .unwrap()
-            .build()
-            .unwrap();
+            )?
+            .limit(3, Some(1))?
+            .build()?;
 
         let expected = "Limit: skip=3, fetch=1\
             \n  Inner Join: Using test.a = test1.a\
             \n    Limit: skip=2, fetch=1\
             \n      TableScan: test\
             \n    TableScan: test1";
-        assert_optimized_plan_eq(&plan, expected);
+        assert_optimized_plan_eq(&plan, expected)
     }
 
     #[test]
-    fn remove_zero_offset() {
-        let table_scan = test_table_scan().unwrap();
+    fn remove_zero_offset() -> Result<()> {
+        let table_scan = test_table_scan()?;
         let plan = LogicalPlanBuilder::from(table_scan)
-            .aggregate(vec![col("a")], vec![sum(col("b"))])
-            .unwrap()
-            .limit(0, None)
-            .unwrap()
-            .build()
-            .unwrap();
+            .aggregate(vec![col("a")], vec![sum(col("b"))])?
+            .limit(0, None)?
+            .build()?;
 
         let expected = "Aggregate: groupBy=[[test.a]], aggr=[[SUM(test.b)]]\
             \n  TableScan: test";
-        assert_optimized_plan_eq(&plan, expected);
+        assert_optimized_plan_eq(&plan, expected)
+    }
+
+    #[test]
+    fn limit_fetch_with_ancestor_limit_skip() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .aggregate(vec![col("a")], vec![sum(col("b"))])?
+            .limit(0, Some(2))?
+            .limit(2, None)?
+            .build()?;
+
+        // No aggregate / scan / limit
+        let expected = "EmptyRelation";
+        assert_optimized_plan_eq_with_pushdown(&plan, expected)
+    }
+
+    #[test]
+    fn limit_with_ancestor_limit() -> Result<()> {
+        let table_scan = test_table_scan()?;
+        let plan = LogicalPlanBuilder::from(table_scan)
+            .aggregate(vec![col("a")], vec![sum(col("b"))])?
+            .limit(2, Some(1))?
+            .sort(vec![col("a")])?
+            .limit(3, Some(1))?
+            .build()?;
+
+        // After remove global-state, we don't record the parent <skip, fetch>
+        // So, bottom don't know parent info, so can't eliminate.
+        let expected = "Limit: skip=3, fetch=1\
+        \n  Sort: test.a, fetch=4\
+        \n    Limit: skip=2, fetch=1\
+        \n      Aggregate: groupBy=[[test.a]], aggr=[[SUM(test.b)]]\
+        \n        TableScan: test";
+        assert_optimized_plan_eq_with_pushdown(&plan, expected)
```

Comment:
Some case need to record information cross plannode.
Exist regression, but it's trivial.

Reply:
Removing the global state does look good in some scenarios, but maybe we should keep it where it's necessary?

Reply:
I think it would make it easier to see the changes in this PR if the order of the tests was the same (not sure if you reordered the tests on purpose or if that is something related to how github is displaying them)